### PR TITLE
pkg/trace/sampler: Fixes Rare Sampler logic

### PR DESCRIPTION
**Pull request by Barbayar**
### What does this PR do?
Fixes Rare Sampler.

### Motivation
The current logic of handling traces without priority looks broken.

Let's say we had multiple top levels in the chunk. And, during the iteration, we decided to sample the trace at `span2`. In this case, only `span2`, `span3`, and `span4` will get refreshed in `seenSpans`. Because of this logic, `span1` could get sampled by Rare Sampler again in a short time, even though we've already sampled a trace with it.
```
span1 <- unsampled
span2 <- sampled
span3
span4
```

### Additional Notes
N/A

### Possible Drawbacks / Trade-offs
- increases the number of iterations over spans
- reduces the number of traces sampled with rare sampler

### Describe how to test/QA your changes
- Number of traces kept with rare sampler should get reduced by a small degree

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/trace/sampler/rare_sampler.go [hunk 1]
```diff
@@ -94,17 +94,19 @@ func (e *RareSampler) handlePriorityTrace(now time.Time, env string, t *pb.Trace
...
 		}
-		if !sampled {
-			sampled = e.sampleSpan(now, env, s)
-			continue
+		if sampled = e.sampleSpan(now, env, s); sampled {
+			break
 		}
-		e.addSpan(expire, env, s)
 	}
+
```
Comment (gbbr) [inline pkg/trace/sampler/rare_sampler.go:105]: Please remove the empty lines below and above the if clause

File: releasenotes/notes/fixes-rare-sampler-7f2167cb7e3337d0.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+fixes:
+  - |
+    Fixed trace rare sampler's oversampling behaviour. With this fix, the rare sampler will sample rare traces more accurately.
```
Comment (gbbr) [inline releasenotes/notes/fixes-rare-sampler-7f2167cb7e3337d0.yaml:11]: We have to add the APM: prefix

**Review by gbbr — APPROVED** [review]
Feel free to merge after addressing the nits.

**Review by raphaelgavache — APPROVED** [review]
nice catch

**Review by dumontg — APPROVED** [review]

**Review by jtappa — CHANGES_REQUESTED** [review]

**Review by jtappa — APPROVED** [review]

**Pull request merged**